src/data/polynomial.lean

```diff
@@ -1365,6 +1403,56 @@ lemma dvd_iff_is_root : (X - C a) ∣ p ↔ is_root p a :=
 lemma mod_by_monic_X (p : polynomial α) : p %ₘ X = C (p.eval 0) :=
 by rw [← mod_by_monic_X_sub_C_eq_C_eval, C_0, sub_zero]
 
+section multiplicity
+
+variable [decidable_rel ((∣) : polynomial α → polynomial α → Prop)]
+
+def root_multiplicity (p : polynomial α) (a : α) : ℕ :=
+if h0 : p = 0 then 0 else
+(multiplicity (X - C a) p).get (multiplicity_finite_of_degree_pos_of_monic
+  (have (0 : α) ≠ 1, from (λ h, by haveI := subsingleton_of_zero_eq_one _ h;
+    exact h0 (subsingleton.elim _ _)),
+  by letI : nonzero_comm_ring α := { zero_ne_one := this, ..show comm_ring α, by apply_instance };
+    rw degree_X_sub_C; exact dec_trivial)
+  (monic_X_sub_C _) h0)
+
+lemma pow_root_multiplicity_dvd (p : polynomial α) (a : α) :
+  (X - C a) ^ root_multiplicity p a ∣ p :=
+if h : p = 0 then by simp [h]
+else by rw [root_multiplicity, dif_neg h];
+  exact multiplicity.pow_multiplicity_dvd _
+
+lemma div_by_monic_mul_pow_root_multiplicity_eq
+  (p : polynomial α) (a : α) :
+  p /ₘ ((X - C a) ^ root_multiplicity p a) *
+  (X - C a) ^ root_multiplicity p a = p :=
+have monic ((X - C a) ^ root_multiplicity p a),
+  from monic_pow (monic_X_sub_C _) _,
+by conv_rhs { rw [← mod_by_monic_add_div p this,
+    (dvd_iff_mod_by_monic_eq_zero this).2 (pow_root_multiplicity_dvd _ _)] };
+  simp [mul_comm]
+
+lemma eval_div_by_monic_pow_root_multiplicity_ne_zero
+  {p : polynomial α} (a : α) (hp : p ≠ 0) :
+  (p /ₘ ((X - C a) ^ root_multiplicity p a)).eval a ≠ 0 :=
+have (0 : α) ≠ 1, from (λ h, by haveI := subsingleton_of_zero_eq_one _ h;
+  exact hp (subsingleton.elim _ _)),
+begin
+  letI : nonzero_comm_ring α := { zero_ne_one := this, ..show comm_ring α, by apply_instance },
```

Comment:
Maybe we could even have a tactic which provided a proof `p ≠ 0` provides us with a local `nonzero_comm_ring` instance?